ouroboros-consensus/src/Ouroboros/Storage/ChainDB/API.hs

```diff
@@ -0,0 +1,330 @@
+      --
+      -- This check is only reliable for blocks up to @k@ away from the tip.
+      -- For blocks older than that the results should be regarded as
+      -- non-deterministic.
+    , getIsFetched       :: STM m (Point blk -> Bool)
+
+      -- | Stream blocks
+      --
+      -- Streaming is not restricted to the current fork, but there must be
+      -- an unbroken path from the starting point to the end point /at the time
+      -- of initialization/ of the iterator. Once the iterator has been
+      -- initialized, it will not be affected by subsequent calls to 'addBlock'.
+      --
+      -- The iterator /does/ have a limited lifetime, however. The chain DB
+      -- internally partitions the chain into an " immutable " part and a
+      -- " volatile " part, moving blocks from the volatile DB to the immutable
+      -- DB when they become more than @k@ deep into the chain. When a block
+      -- with slot number @n@ is added to the immutble DB, a time delay @t@
+      -- kicks in; after that time delay expires, all blocks older than @n@ may
+      -- be removed from the volatile DB, /including any blocks that happen to
+      -- live on other forks/ (since those forks must now, by definition, be too
+      -- distant). This time delay @t@ also provides a worst-case bound for the
+      -- lifetime of the iterator: if the iterator traverses a chain that
+      -- forks off from our current chain at the tip of the immutable DB,
+      -- then the first block on that fork will become unavailable as soon as
+      -- another block is pushed to the current chain and the subsequent
+      -- time delay expires.
+    , streamBlocks       :: StreamFrom blk -> StreamTo blk -> m (Iterator m blk)
+
+      -- | Chain reader
+      --
+      -- A chain reader is an iterator that tracks the state of the /current/
+      -- chain: calling @next@ on the iterator will either give you the next
+      -- block, or (if we have switched to a fork) the instruction to rollback.
+      --
+      -- The tracking iterator starts at genesis (see also 'trackForward').
+    , newReader          :: m (Reader m hdr)
+
+      -- | Known to be invalid blocks
+    , knownInvalidBlocks :: STM m (Set (Point blk))
+
+      -- | Check if the specified point is on the current chain
+      --
+      -- This lives in @m@, not @STM m@, because if the point is not on the
+      -- current chain fragment, it might have to query the immutable DB.
+    , pointOnChain       :: Point blk -> m Bool
+    }
+
+{-------------------------------------------------------------------------------
+  Derived
+-------------------------------------------------------------------------------}
+
+getTipHeader :: MonadSTM m => ChainDB m blk hdr -> STM m hdr
+getTipHeader ChainDB{..} =
+    mustExist . Fragment.head <$> getCurrentChain
+  where
+    mustExist :: Maybe a -> a
+    mustExist (Just a) = a
+    mustExist Nothing  = error "invariant violation: empty current chain"
```

Comment:
Yes, fair point. This is a remnant of when I was going to do the unthinkable and take _advantage_ of EBBs :flushed: Let me quickly hide all evidence of that momentary lapse in sanity.

Reply:
There is a second problem here too, which is that the current chain fragment may in fact be empty even if the immutable DB is not (if for some reason we lost the entire volatile DB).

Reply:
I'll move this into the record itself, but I think the type will need to change, and will need to become `m`, not `STM m`, because if this is the immutable DB, we'd have to read it from disk? I don't know if that will cause trouble in the ChainDB clients... :thinking:

Reply:
I see. Currently, I'm not using this function anyway so I suppose it will be fine.